File: live2d/companion.js
/* Aster Companion API v1. Original adapter code; contains no model or SDK assets.
 * Browser: AsterCompanion.create(coreModel, profile). Node: require this file.
 * Profiles are data. No eval, callbacks, resource URLs or expression execution.
 */
(function (root) {
  'use strict';
  const STATES = ['idle','listening','thinking','reading','working','checking','waiting','speaking','pleased','concerned'];
  const clone = value => JSON.parse(JSON.stringify(value));
  const own = (object, key) => Object.prototype.hasOwnProperty.call(object, key);
  const record = value => value !== null && typeof value === 'object' && !Array.isArray(value);
  const name = value => typeof value === 'string' && /^[A-Za-z0-9_-]{1,64}$/.test(value);
  const finite = (value, min, max) => typeof value === 'number' && Number.isFinite(value) && value >= min && value <= max;
  const fail = message => { throw new Error(message); };
  function keys(value, allowed, required = allowed) {
    if (!record(value) || Object.keys(value).some(k => !allowed.includes(k)) || required.some(k => !own(value,k))) fail('Unknown or missing fields');
  }
  function validateProfile(p) {
    if (new TextEncoder().encode(JSON.stringify(p)).length > 128000) fail('Profile exceeds 128 KB');
    keys(p,['version','name','model','layout','bindings','emotions','motions']);
    if (p.version !== 1 || typeof p.name !== 'string' || !p.name || new TextEncoder().encode(p.name).length > 120 || /[\x00-\x1f\x7f]/.test(p.name)) fail('Invalid profile version or name');
    if (typeof p.model !== 'string' || new TextEncoder().encode(p.model).length > 512 || !p.model.endsWith('.model3.json') || /[\\:?#%\x00-\x1f\x7f]/.test(p.model) || p.model.split('/').some(s => !s || s === '.' || s === '..')) fail('Use a relative local model path');
    keys(p.layout,['zoom','center','anchor']);
    if (!finite(p.layout.zoom,.1,5) || ['center','anchor'].some(k => !Array.isArray(p.layout[k]) || p.layout[k].length !== 2 || p.layout[k].some(n => !finite(n,0,1)))) fail('Invalid layout');
    if (!record(p.bindings) || !Object.keys(p.bindings).length || Object.keys(p.bindings).length > 128 || Object.entries(p.bindings).some(([k,v]) => !name(k) || !name(v)) || new Set(Object.values(p.bindings)).size !== Object.keys(p.bindings).length) fail('Invalid or duplicate parameter bindings');
    if (!record(p.emotions) || !own(p.emotions,'neutral') || Object.keys(p.emotions).length > 32 || !record(p.motions) || Object.keys(p.motions).length > 32) fail('Define neutral and at most 32 emotions/motions');
    for (const [n, values] of Object.entries(p.emotions)) {
      if (!name(n) || !record(values) || Object.entries(values).some(([k,v]) => !own(p.bindings,k) || !finite(v,-10000,10000))) fail('Invalid emotion: '+n);
    }
    for (const [n, clip] of Object.entries(p.motions)) {
      keys(clip,['duration_ms','blend','tracks']);
      if (!name(n) || !Number.isInteger(clip.duration_ms) || !finite(clip.duration_ms,100,10000) || !['add','set'].includes(clip.blend) || !record(clip.tracks) || !Object.keys(clip.tracks).length || Object.keys(clip.tracks).length > 128) fail('Invalid motion: '+n);
      for (const [channel, frames] of Object.entries(clip.tracks)) {
        if (!own(p.bindings,channel) || !Array.isArray(frames) || frames.length < 2 || frames.length > 64 || frames.some((f,i) => !Array.isArray(f) || f.length !== 2 || !finite(f[0],0,1) || !finite(f[1],-10000,10000) || (i > 0 && frames[i-1][0] >= f[0])) || frames[0][0] !== 0 || frames.at(-1)[0] !== 1) fail('Invalid keyframe track: '+n+'/'+channel);
      }
    }
    return clone(p);
  }
  function create(core, source) {
    const p = validateProfile(source);
    const raw = core.getModel().parameters;
    const parameters = Array.from(raw.ids, (id,i) => ({id:String(id), min:raw.minimumValues[i], max:raw.maximumValues[i], default:raw.defaultValues[i], index:i}));
    if (parameters.length > 4096 || parameters.some(v => !finite(v.min,-1e10,1e10) || !finite(v.max,v.min,1e10) || !finite(v.default,v.min,v.max))) fail('Invalid model parameter metadata');
    const byId = new Map(parameters.map(v => [v.id,v]));
    const bindings = new Map(Object.entries(p.bindings).filter(([,id]) => byId.has(id)));
    const warnings = Object.entries(p.bindings).filter(([,id]) => !byId.has(id)).map(([channel,id]) => 'Unsupported binding '+channel+': '+id);
    let time = 0, state = 'idle', emotion = 'neutral', intensity = 1, clip = null, gaze = null, last = {};
    const supported = channels => channels.filter(channel => bindings.has(channel));
    const description = () => ({version:1, profile:p.name, model:p.model, states:STATES.slice(), emotions:Object.keys(p.emotions), motions:Object.keys(p.motions), bindings:clone(p.bindings), parameters:parameters.map(({index,...v}) => v), warnings:warnings.slice()});
    function setState(value) { if (!STATES.includes(value)) fail('Unknown work state: '+value); state = value; }
    function strength(value) { if (!finite(value,0,1)) fail('Strength must be 0–1'); return value; }
    function setEmotion(value, weight = 1) {
      if (!own(p.emotions,value)) fail('Unknown emotion: '+value);
      strength(weight);
      const channels = supported(Object.keys(p.emotions[value]));
      if (Object.keys(p.emotions[value]).length && !channels.length) fail('Emotion has no supported model parameters');
      emotion = value; intensity = weight;
      return {emotion:value, strength:weight, supported_channels:channels};
    }
    function play(value, weight = 1) {
      if (!own(p.motions,value)) fail('Unknown motion: '+value);
      strength(weight);
      const channels = supported(Object.keys(p.motions[value].tracks));
      if (!channels.length) fail('Motion has no supported model parameters');
      clip = {name:value, start:time, strength:weight};
      return {motion:value, duration_ms:p.motions[value].duration_ms, supported_channels:channels};
    }
    function look(x,y,duration_ms = 2000) {
      if (!finite(x,-1,1) || !finite(y,-1,1) || !Number.isInteger(duration_ms) || !finite(duration_ms,100,10000)) fail('Invalid gaze coordinates or duration');
      const channels = supported(['eye_x','eye_y','head_x','head_y']);
      if (!channels.length) fail('Gaze has no supported model parameters');
      gaze = {x,y,start:time,end:time+duration_ms/1000};
      return {look:[x,y], duration_ms, supported_channels:channels};
    }
    function reset() { emotion='neutral'; intensity=1; clip=null; gaze=null; return {reset:true}; }
    function dispatch(command) {
      if (!record(command)) fail('Expected a command object');
      switch (command.type) {
        case 'emotion': keys(command,['type','name','strength']); return setEmotion(command.name,command.strength);
        case 'motion': keys(command,['type','name','strength']); return play(command.name,command.strength);
        case 'look': keys(command,['type','x','y','duration_ms']); return look(command.x,command.y,command.duration_ms);
        case 'reset': keys(command,['type']); return reset();
        default: fail('Unknown companion command');
      }
    }
    function interpolate(frames, at) {
      const end = frames.findIndex(f => f[0] >= at);
      if (end <= 0) return frames[0][1];
      const a=frames[end-1], b=frames[end], ratio=(at-a[0])/(b[0]-a[0]);
      return a[1]+(b[1]-a[1])*ratio;
    }
    function apply(delta_ms = 100) {
      if (!finite(delta_ms,0,250)) fail('Frame delta must be 0–250 ms');
      time += delta_ms/1000;
      const values = Object.fromEntries(Array.from(bindings, ([channel,id]) => [channel, byId.get(id).default]));
      const put = (key,value) => { if (bindings.has(key)) values[key]=value; };
      const blinkPhase=time%4.7, blink=blinkPhase>4.48?Math.abs((blinkPhase-4.59)/.11):1;
      const thinking=['thinking','reading','checking'].includes(state), attentive=['listening','waiting'].includes(state), sway=Math.sin(time*.8)*.8;
      put('head_x',sway+(thinking?-6:0)+(state==='working'?-3:0));
      put('head_y',Math.sin(time*.6)*.6+(attentive?2:0)); put('head_z',thinking?-2:0); put('body_x',sway*.3);
      put('eye_x',thinking?-.3:0); put('eye_y',state==='reading'?-.18:(thinking?.12:0));
      put('eye_l',Math.max(0,Math.min(state==='concerned'?.7:1,blink))); put('eye_r',Math.max(0,Math.min(state==='concerned'?.7:1,blink)));
      put('breath',.5+Math.sin(time*1.3)*.5); put('mouth_open',state==='speaking'?Math.max(0,Math.sin(time*15)*.65+.25):0);
      put('smile_l',state==='pleased'?.45:0); put('smile_r',state==='pleased'?.45:0);
      for (const [channel,value] of Object.entries(p.emotions.neutral)) put(channel,value);
      for (const [channel,value] of Object.entries(p.emotions[emotion])) if (bindings.has(channel)) put(channel,values[channel]+(value-values[channel])*intensity);
      if (gaze && time >= gaze.end) gaze=null;
      if (gaze) {
        const blend=Math.min(1,(time-gaze.start)/.15,(gaze.end-time)/.15);
        for (const [channel,value] of Object.entries({eye_x:gaze.x,eye_y:gaze.y,head_x:gaze.x*20,head_y:gaze.y*10})) if(bindings.has(channel)) put(channel,values[channel]+(value-values[channel])*blend);
      }
      if (clip) {
        const definition=p.motions[clip.name], elapsed=time-clip.start, duration=definition.duration_ms/1000;
        if (elapsed >= duration) clip=null;
        else {
          const weight=clip.strength*Math.min(1,elapsed/.1,(duration-elapsed)/.1);
          for (const [channel,frames] of Object.entries(definition.tracks)) if (bindings.has(channel)) {
            const value=interpolate(frames,elapsed/duration);
            values[channel] = definition.blend==='add' ? values[channel]+value*weight : values[channel]+(value-values[channel])*weight;
          }
        }
      }
      last={};
      for (const [channel,id] of bindings) {
        const meta=byId.get(id), value=Math.max(meta.min,Math.min(meta.max,values[channel]));
        core.setParameterValueByIndex(meta.index,value); last[id]=value;
      }
      return clone(last);
    }
    const snapshot = () => ({version:1,state,emotion,strength:intensity,motion:clip?.name||null,look:gaze?[gaze.x,gaze.y]:null,time,parameters:clone(last)});
    return Object.freeze({describe:description,state:setState,emotion:setEmotion,motion:play,look,reset,dispatch,apply,snapshot});
  }
  const api=Object.freeze({version:1,validateProfile,create});
  if (typeof module !== 'undefined' && module.exports) module.exports=api;
  else root.AsterCompanion=api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: src/companion.rs
//! Public, declarative companion profiles. No executable code or model assets.
use anyhow::{Context, Result, bail};
use serde::{Deserialize, Serialize};
use std::{collections::BTreeMap, fs, io::Read, path::Path};

pub const DEFAULT_PROFILE: &str = include_str!("../live2d/profiles/nongyu.json");
pub const STATES: &[&str] = &[
    "idle",
    "listening",
    "thinking",
    "reading",
    "working",
    "checking",
    "waiting",
    "speaking",
    "pleased",
    "concerned",
];
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Layout {
    pub zoom: f64,
    pub center: [f64; 2],
    pub anchor: [f64; 2],
}
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Clip {
    pub duration_ms: u32,
    pub blend: String,
    pub tracks: BTreeMap<String, Vec<[f64; 2]>>,
}
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Profile {
    pub version: u32,
    pub name: String,
    pub model: String,
    pub layout: Layout,
    pub bindings: BTreeMap<String, String>,
    pub emotions: BTreeMap<String, BTreeMap<String, f64>>,
    pub motions: BTreeMap<String, Clip>,
}
pub fn relative(path: &str) -> bool {
    !path.is_empty()
        && path.len() <= 512
        && !path.starts_with('/')
        && !path.contains(['\\', ':', '?', '#', '%'])
        && !path.chars().any(char::is_control)
        && path.split('/').all(|part| !matches!(part, "" | "." | ".."))
}
fn name(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 64
        && value
            .bytes()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, b'_' | b'-'))
}
impl Profile {
    pub fn load(path: Option<&Path>) -> Result<Self> {
        let text = if let Some(path) = path {
            let mut text = String::new();
            fs::File::open(path)?
                .take(128_001)
                .read_to_string(&mut text)?;
            if text.len() > 128_000 {
                bail!("Companion profile exceeds 128 KB")
            }
            text
        } else {
            DEFAULT_PROFILE.into()
        };
        let profile: Self =
            serde_json::from_str(&text).context("Invalid companion profile JSON")?;
        profile.validate()?;
        Ok(profile)
    }
    pub fn validate(&self) -> Result<()> {
        if self.version != 1
            || self.name.is_empty()
            || self.name.len() > 120
            || self.name.chars().any(char::is_control)
            || !relative(&self.model)
            || !self.model.ends_with(".model3.json")
        {
            bail!(
                "Use companion version 1, a name up to 120 bytes, and a relative .model3.json path"
            );
        }
        if !self.layout.zoom.is_finite()
            || !(0.1..=5.0).contains(&self.layout.zoom)
            || self
                .layout
                .center
                .iter()
                .chain(self.layout.anchor.iter())
                .any(|n| !n.is_finite() || !(0.0..=1.0).contains(n))
        {
            bail!("Layout zoom must be 0.1–5; center and anchor coordinates must be 0–1");
        }
        if self.bindings.is_empty()
            || self.bindings.len() > 128
            || self.bindings.iter().any(|(k, v)| !name(k) || !name(v))
            || self
                .bindings
                .values()
                .collect::<std::collections::BTreeSet<_>>()
                .len()
                != self.bindings.len()
        {
            bail!("Use 1–128 unique parameter bindings with ASCII names up to 64 bytes");
        }
        if !self.emotions.contains_key("neutral")
            || self.emotions.len() > 32
            || self.motions.len() > 32
        {
            bail!("Define neutral and at most 32 emotions and 32 motions");
        }
        let valid_value = |v: &f64| v.is_finite() && v.abs() <= 10_000.0;
        for (emotion, values) in &self.emotions {
            if !name(emotion)
                || values
                    .iter()
                    .any(|(key, value)| !self.bindings.contains_key(key) || !valid_value(value))
            {
                bail!(
                    "Invalid emotion {emotion}: use bound channels and finite values within ±10000"
                );
            }
        }
        for (motion, clip) in &self.motions {
            if !name(motion)
                || !(100..=10_000).contains(&clip.duration_ms)
                || !matches!(clip.blend.as_str(), "add" | "set")
                || clip.tracks.is_empty()
                || clip.tracks.len() > 128
            {
                bail!(
                    "Invalid motion {motion}: use 100–10000 ms, add/set blending and 1–128 tracks"
                );
            }
            for (channel, frames) in &clip.tracks {
                if !self.bindings.contains_key(channel)
                    || !(2..=64).contains(&frames.len())
                    || frames[0][0] != 0.0
                    || frames.last().unwrap()[0] != 1.0
                    || frames.iter().any(|[at, v]| {
                        !at.is_finite() || !(0.0..=1.0).contains(at) || !valid_value(v)
                    })
                    || frames.windows(2).any(|pair| pair[0][0] >= pair[1][0])
                {
                    bail!(
                        "Invalid track {motion}/{channel}: use 2–64 ordered [time,value] frames from time 0 to 1"
                    );
                }
            }
        }
        Ok(())
    }
}
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum Control {
    Emotion { name: String, strength: f64 },
    Motion { name: String, strength: f64 },
    Look { x: f64, y: f64, duration_ms: u32 },
    Reset,
}
impl Control {
    pub fn validate(&self) -> Result<()> {
        match self {
            Self::Emotion { name: n, strength } | Self::Motion { name: n, strength } => {
                if !name(n) || !strength.is_finite() || !(0.0..=1.0).contains(strength) {
                    bail!("Use a profile name and strength from 0 to 1")
                }
            }
            Self::Look { x, y, duration_ms } => {
                if !x.is_finite()
                    || !y.is_finite()
                    || x.abs() > 1.0
                    || y.abs() > 1.0
                    || !(100..=10000).contains(duration_ms)
                {
                    bail!("Gaze coordinates must be −1 to 1, duration 100–10000 ms")
                }
            }
            Self::Reset => {}
        }
        Ok(())
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn profiles_and_controls_reject_unsafe_or_unbounded_input() {
        let mut p = Profile::load(None).unwrap();
        assert_eq!(p.version, 1);
        p.model = "../private.model3.json".into();
        assert!(p.validate().is_err());
        p = Profile::load(None).unwrap();
        p.motions
            .get_mut("nod")
            .unwrap()
            .tracks
            .get_mut("head_y")
            .unwrap()[1][0] = 0.0;
        assert!(p.validate().is_err());
        assert!(
            Control::Emotion {
                name: "happy".into(),
                strength: f64::NAN
            }
            .validate()
            .is_err()
        );
        assert!(
            Control::Look {
                x: 2.0,
                y: 0.0,
                duration_ms: 1000
            }
            .validate()
            .is_err()
        );
        assert!(!relative("https://example.com/rig.model3.json"));
        assert!(!relative("model/%2e%2e/file"));
        assert!(relative("my pet/rig.model3.json"));
        let extra = DEFAULT_PROFILE.replacen(
            "\"version\": 1",
            "\"version\": 1, \"javascript\": \"execute()\"",
            1,
        );
        assert!(serde_json::from_str::<Profile>(&extra).is_err());
    }
}
